src/coreclr/src/vm/ipcstreamfactory.h

```diff
@@ -0,0 +1,101 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef __IPC_STREAM_FACTORY_H__
+#define __IPC_STREAM_FACTORY_H__
+
+#ifdef FEATURE_PERFTRACING
+
+#include "diagnosticsipc.h"
+
+class IpcStreamFactory
+{
+public:
+    struct ConnectionState
+    {
+    public:
+        ConnectionState(IpcStream::DiagnosticsIpc *pIpc) :
+            _pIpc(pIpc),
+            _pStream(nullptr)
+        { }
+
+        // returns a pollable handle and performs any preparation required
+        // e.g., as a side-effect, will connect and advertise on reverse connections
+        virtual bool GetIpcPollHandle(IpcStream::DiagnosticsIpc::IpcPollHandle *pIpcPollHandle, ErrorCallback callback = nullptr) = 0;
+
+        // Returns the signaled stream in a usable state
+        virtual IpcStream *GetConnectedStream(ErrorCallback callback = nullptr) = 0;
+
+        // Resets the connection in the event of a hangup
+        virtual void Reset(ErrorCallback callback = nullptr) = 0;
+
+        // closes the underlying connections
+        void Close(ErrorCallback callback = nullptr)
+        {
+            if (_pIpc != nullptr)
+                _pIpc->Close(callback);
+            if (_pStream != nullptr)
+                _pStream->Close(callback);
+        }
```

Comment:
_pIpc and _pStream need to be deleted (not that a leak on shutdown would really matter, but may as well clean up properly)

Reply:
I specifically leaked these since the PAL is unaware of shutdown paths and has access to these pointers.  This method is specifically called from a different thread than the server thread.  If we delete them in this code path it can lead to AVs on the server thread in the PAL (inside `IpcStream::DiagnosticsIpc::Poll` specifically).  Rather than introduce some form of knowledge of runtime state to the PAL or potentially adding locks protecting access I opted to just leak the memory on shutdown, since that is the only place this method is called.  I should add a comment that calls out this assumption, though.